Properly allow macro expanded `format_args` invocations to uses captures

Noratrieb opened the pull request:
Originally, this was kinda half-allowed. There were some primitive checks in place that looked at the span to see whether the input was likely a literal. These "source literal" checks are needed because the spans created during `format_args` parsing only make sense when it is indeed a literal that was written in the source code directly.

This is orthogonal to the restriction that the first argument must be a "direct literal", not being exanpanded from macros. This restriction was imposed by [RFC 2795] on the basis of being too confusing. But this was only concerned with the argument of the invocation being a literal, not whether it was a source literal (maybe in spirit it meant it being a source literal, this is not clear to me).

Since the original check only really cared about source literals (which is good enough to deny the `format_args!(concat!())` example), macros expanding to `format_args` invocations were able to use implicit captures if they spanned the string in a way that lead back to a source string.

The "source literal" checks were not strict enough and caused ICEs in certain cases (see #106191). So I tightened it up in #106195 to really only work if it's a direct source literal.

This caused the `indoc` crate to break. `indoc` transformed the source literal by removing whitespace, which made it not a "source literal" anymore (which is required to fix the ICE). But since `indoc` spanned the literal in ways that made the old check think that it's a literal, it was able to use implicit captures (which is useful and nice for the users of `indoc`).

This commit properly seperates the previously introduced concepts of "source literal" and "direct literal" and therefore allows `indoc` invocations, which don't create "source literals" to use implicit captures again.

Fixes #106191

[RFC 2795]: https://rust-lang.github.io/rfcs/2795-format-args-implicit-identifiers.html#macro-hygiene

Noratrieb committed:
Properly allow macro expanded `format_args` invocations to uses captures

Originally, this was kinda half-allowed. There were some primitive
checks in place that looked at the span to see whether the input was
likely a literal. These "source literal" checks are needed because the
spans created during `format_args` parsing only make sense when it is
indeed a literal that was written in the source code directly.

This is orthogonal to the restriction that the first argument must be a
"direct literal", not being exanpanded from macros. This restriction was
imposed by [RFC 2795] on the basis of being too confusing. But this was
only concerned with the argument of the invocation being a literal, not
whether it was a source literal (maybe in spirit it meant it being a
source literal, this is not clear to me).

Since the original check only really cared about source literals (which
is good enough to deny the `format_args!(concat!())` example), macros
expanding to `format_args` invocations were able to use implicit
captures if they spanned the string in a way that lead back to a source
string.

The "source literal" checks were not strict enough and caused ICEs in
certain cases (see # 106191 (the space is intended to avoid spammy
backreferences)). So I tightened it up in # 106195 to really only work
if it's a direct source literal.

This caused the `indoc` crate to break. `indoc` transformed the source
literal by removing whitespace, which made it not a "source literal"
anymore (which is required to fix the ICE). But since `indoc` spanned
the literal in ways that made the old check think that it's a literal,
it was able to use implicit captures (which is useful and nice for the
users of `indoc`).

This commit properly seperates the previously introduced concepts of
"source literal" and "direct literal" and therefore allows `indoc`
invocations, which don't create "source literals" to use implicit
captures again.

[RFC 2795]: https://rust-lang.github.io/rfcs/2795-format-args-implicit-identifiers.html#macro-hygiene

Nilstrieb commented:
Since this changes what kind of code is allowed to compile and settles a question not fully answered by the RFC, I'll lang nominate this.

# Question for the lang team

*See the context in the PR description first.*

Should `format_args!` invocations that were generated by a proc-macro be able to use implicit argument captures?

A concrete example that does not compile before this PR (neither on stable nor nightly) but does compile with it:
`pm.rs`
```rust
extern crate proc_macro;

#[proc_macro]
pub fn foo(_: proc_macro::TokenStream) -> proc_macro::TokenStream {
    r#"{ let x = 5; format!("{x}") }"#.parse().unwrap()
}
```
`rustc pm.rs --crate-type proc-macro`
```main.rs
fn main() {
    dbg!(pm::foo!());
}
```
`rustc main.rs --extern pm=libpm.so`

## Arguments for it

The `indoc` crate is an example of a crate that makes usage of this. Note that format args captures `indoc` compiles on stable but not nightly because of the old broken and inconsistent semantics. It's in my opinion unreasonable to support the `indoc` usage but not my example above (since only the spans of the string literal are different).

## Arguments against it

This could lead to confusing behavior in contrived situations. `indoc` (or other crates like it) could do their own parsing of the implicit captures and expand them out into separate bindings before forwarding them to `format_args`, but this can of course be said for the whole `format_args` machinery.

tmandry commented:
Discussed in yesterday's lang team meeting.

> Should `format_args!` invocations that were generated by a proc-macro be able to use implicit argument captures?

Several of us had the instinct that the answer should be "yes, if the span attached to the `{foo}` has hygiene that can name `foo`". It's possible that this answer misses some nuance, but if it does we're missing that context.

> It's in my opinion unreasonable to support the indoc usage but not my example above (since only the spans of the string literal are different).

I think I agree, but I don't have context on what the spans of the string literal produced by indoc are. I would guess that it uses `Span::call_site()`?

> `indoc` (or other crates like it) could do their own parsing of the implicit captures and expand them out into separate bindings before forwarding them to `format_args`, but this can of course be said for the whole `format_args` machinery.

I don't follow. Can you please give an example?

Generally, we'd like to see a comment that talks about the behavior in the RFC (to the extent that it's specified) and compares it to the old behavior, current behavior, and behavior after landing this PR.

In the short term, we think it makes sense to revert https://github.com/rust-lang/rust/pull/106195 (and backport that revert if needed) so that we don't break existing code in the next release. Fixing an ICE is lower priority than keeping existing code working. Then we can discuss what the right behavior is and implement that without the time pressure.

Please re-nominate for discussion when our input is needed again. Thanks!

scottmcm commented:
We discussed this in the lang team meeting today.  Broadly we thought that proc macros being able to emit tokens that use captures makes logical sense to me.  So I'll start a

@rfcbot fcp merge

but we had some requests to confirm, so...

scottmcm commented:
We discussed this in the lang team meeting today. Broadly we thought that proc macros being able to emit tokens that use captures makes logical sense to me. So I'll start a

@rfcbot fcp merge

but we had some requests to confirm, so...

scottmcm commented:
@rfcbot concern please-link-tests

We did discuss a bunch of situations in the meeting where we'd like to get them listed out to affirmatively confirm those cases.  It sounds like a bunch of them already work as expected, but we'd like to have tests (or pointers to existing tests) for things like how hygiene works in macro-generated format strings with captures, which things don't work, how it relates to when things are expanded, etc.  (Like is `format!(macro_be!())` different from `format!(p_macro!())`?)

petrochenkov commented:
>a proc macro *emitting* something as a whole that one could write directly ... seems entirely fine, though mixing things like `format!(concat!(…))` might still be best to reject

This was always my understanding as well.
We parse the expression at the start of `TOKENS` in `format_args!(TOKENS)`, if it's a literal - good, if it's not a literal - bad.
It's not clear why the implementation considered spans at all.

petrochenkov commented:
The current implementation seems overcomplicated too - we don't need `is_string_literal_including_nt`, we only need to parse an expression and then do a post factum check on it.

tmandry commented:
> `{ let x = 5; format!("{x}") }` (no macro input, any hygiene for `x`)

This PR still requires the hygiene for `x` in the binding and format string to match, right?

Nilstrieb commented:
Yes. Well, it doesn't require them to match for macro expansion to succeed, but x will not be found later if it doesn't match.

Nilstrieb commented:
I have collected and added a few tests:

`tests/ui/fmt/indoc-issue-106408.rs` tests that hygiene info is preserved correctly through a proc macro invocation
`tests/ui/fmt/format-args-capture-first-literal-is-macro.rs`  tests that MBEs, proc macros and builtin macros in argument position block captures
`tests/ui/fmt/format-args-capture-from-pm-first-arg-macro.rs` tests that we still forbid `format!(concat!("{a}"))` even if it comes from a proc macro

joshtriplett commented:
@scottmcm Do the latest changes by @Nilstrieb address your concern?

scottmcm commented:
Thanks, @Nilstrieb !  That's very helpful
@rfcbot resolve please-link-tests

The only other thing that comes to mind is something like this
```
macro_rules! format_mbe {
    ($tt:tt) => {
        {
            let a = 123;
            format!($tt)
        }
    };
}

fn main() {
    let a = 0;
    assert_eq!(format_mbe!("{a}"), "0");
}
```
but if that's not actually different because of how resolution works, then feel free to omit it.

bors commented:
:umbrella: The latest upstream changes (presumably #107788) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Nilstrieb commented:
That also sounds like a useful test to have, so I added it (and rebased).

nikomatsakis commented:
@rfcbot fcp reviewed

petrochenkov commented on compiler/rustc_builtin_macros/src/format.rs:
```diff
@@ -59,25 +70,21 @@ fn parse_args<'a>(
 
+    // Only allow implicit captures to be used when the argument is a direct literal
+    // instead of a macro expanding to one.
+    let is_direct_literal = matches!(fmtstr.kind, ExprKind::Lit(lit) if lit.is_string());
```
```suggestion
    let is_direct_literal = matches!(fmtstr.kind, ExprKind::Lit(_));
```
Does it matter whether the literal is a string?
If it's not then we'll report an error anyway, this property seems orthogonal to that fact.

Nilstrieb replied:
I don't think it matters, yeah.

Nilstrieb commented on compiler/rustc_ast/src/token.rs:
```diff
@@ -103,6 +103,10 @@ impl Lit {
 
+    pub fn is_string(self) -> bool {
+        matches!(self.kind, Str | StrRaw(_))
+    }
```
I just removed it, since one of the uses wasn't needed anyways as you noticed below.

Noratrieb committed:
Improve heuristics for format_args literal being suggestable

Sometimes, we want to create subspans and point at code in the literal
if possible. But this doesn't always make sense, sometimes the literal
may come from macro expanded code and isn't actually there in the
source. Then, we can't really make these suggestions.

This now makes sure that the literal is actually there as we see it so
that we will not run into ICEs on weird literal transformations.

Nilstrieb commented on compiler/rustc_parse_format/src/lib.rs:
```diff
@@ -902,8 +906,27 @@ fn find_width_map_from_snippet(
+    // since we will never need to point our spans there, so we lie about it here by ignoring it.
+    // Since there might actually be newlines in the source code, we need to normalize away all trailing newlines.
+    // If we only trimmed it off the input, `format!("\n")` would cause a mismatch as here we they actually match up.
+    // Alternatively, we could just count the trailing newlines and only trim one from the input if they don't match up.
```
I moved it into a separate commit.

petrochenkov commented on compiler/rustc_parse_format/src/lib.rs:
```diff
@@ -986,6 +1009,19 @@ fn find_width_map_from_snippet(
     InputStringKind::Literal { width_mappings }
 }
 
+fn unescape_string(string: &str) -> Result<string::String, unescape::EscapeError> {
```
```suggestion
fn unescape_string(string: &str) -> Option<string::String> {
```
The error is never used.

Maybe the whole function should also be inlined and removed, not sure.

petrochenkov commented on compiler/rustc_parse_format/src/lib.rs:
```diff
@@ -986,6 +1009,19 @@ fn find_width_map_from_snippet(
 
+fn unescape_string(string: &str) -> Result<string::String, unescape::EscapeError> {
+    let mut buf = string::String::new();
+    let mut error = Ok(());
```
```suggestion
    let mut ok = true;
```
and `ok.then(|| buf)` in the end.

petrochenkov commented:
r=me after addressing the remaining comments.
@rustbot author

Nilstrieb commented:
@bors r=petrochenkov

bors commented:
:pushpin: Commit 427aceb9d4f022af56ed39163510185d4251dcde has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov 427aceb9d4f022af56ed39163510185d4251dcde -->
<!-- homu: {"type":"Approved","sha":"427aceb9d4f022af56ed39163510185d4251dcde","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 427aceb9d4f022af56ed39163510185d4251dcde with merge 2e7034ebf7f57066c260b680c5e9dfcf04ca4cd5...
<!-- homu: {"type":"BuildStarted","head_sha":"427aceb9d4f022af56ed39163510185d4251dcde","merge_sha":"2e7034ebf7f57066c260b680c5e9dfcf04ca4cd5"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4416782048/jobs/7745994579)
Approved by: petrochenkov
Pushing 2e7034ebf7f57066c260b680c5e9dfcf04ca4cd5 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4416782048/jobs/7745994579"},"merge_sha":"2e7034ebf7f57066c260b680c5e9dfcf04ca4cd5"} -->

rust-timer commented:
Finished benchmarking commit (2e7034ebf7f57066c260b680c5e9dfcf04ca4cd5): [comparison URL](https://perf.rust-lang.org/compare.html?start=669e75163957f8f2408d515ce2da3516cb31f747&end=2e7034ebf7f57066c260b680c5e9dfcf04ca4cd5&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=669e75163957f8f2408d515ce2da3516cb31f747&end=2e7034ebf7f57066c260b680c5e9dfcf04ca4cd5&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 0.4% | [0.4%, 0.5%] | 5     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=669e75163957f8f2408d515ce2da3516cb31f747&end=2e7034ebf7f57066c260b680c5e9dfcf04ca4cd5&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.5%  | [3.5%, 3.5%]   | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.5% | [-3.5%, -3.5%] | 1     |
| All ❌✅ (primary)                 | 3.5%  | [3.5%, 3.5%]   | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=669e75163957f8f2408d515ce2da3516cb31f747&end=2e7034ebf7f57066c260b680c5e9dfcf04ca4cd5&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 4.9%  | [4.1%, 5.5%]   | 4     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.2% | [-2.3%, -2.1%] | 2     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>
